Python/chapter03/p01_three_in_one/miguelHx.py

```diff
@@ -0,0 +1,420 @@
+                'end': stack_capacity * 2 - 1,
+                'size': 0,
+                'top_index': stack_capacity,
+                'top_index_next': -1,
+            },
+            3: {
+                'start': stack_capacity * 2,
+                'end': stack_capacity * 3 - 1,
+                'size': 0,
+                'top_index': stack_capacity * 2,
+                'top_index_next': -1
+            }
+        }
+        self.values = [0] * (stack_capacity * self.num_stacks)
+
+    def is_empty(self, stack_id):
+        return self.stack_info[stack_id]['size'] <= 0
+    
+    def peek(self, stack_id):
+        # this method returns the value at the top index
+        if self.is_empty(stack_id):
+            raise IndexError('Stack is empty. Stack ID: {}'.format(stack_id))
+        return self.values[self.stack_info[stack_id]['top_index']]
+
+    def push(self, stack_id, value):
+        # first, check if stack of interest is full
+        if self.stack_info[stack_id]['size'] >= self.stack_capacity:
+            raise IndexError('Stack is full. Stack ID: {}'.format(stack_id))
+
+        
+
+        stack_top_index = self.stack_info[stack_id]['top_index']
+        new_stack_top_index = stack_top_index + 1
+        # if empty, then top index and next index stay the same.
+        if self.is_empty(stack_id):
+            self.values[stack_top_index] = value
+            # update stack size.
+            self.stack_info[stack_id]['size'] += 1
+            return
+
+        # otherwise, stack is not empty, and so we
+        # first need to update the next and top indices, then update
+        # the values accordingly.
+        self.stack_info[stack_id]['top_index_next'] = stack_top_index
+        self.stack_info[stack_id]['top_index'] = new_stack_top_index
+        self.values[new_stack_top_index] = value
+        self.stack_info[stack_id]['size'] += 1
+        
+        
+    
+    def pop(self, stack_id):
+        # first, make sure we are not at an empty stack
+        if self.is_empty(stack_id):
+            raise IndexError('Stack is empty. Stack ID: {}'.format(stack_id))
+        # if there is one element, top index will be 0
+        # and next index will be -1
+        # if there are two elements, top index will be 1,
+        # and next index will be 0
+        # if there are three elements, top index will be 2,
+        # next will be 1
+        # basically, the next index will be one less than top index
+
+        # when we pop, top index will become next, and next will get
+        # decremented by one. (Or, each get decremented by 1)
+
+        # get stack top index, then set value to 0 as a way of clearing
+        original_stack_top_index = self.stack_info[stack_id]['top_index']
+        val_before_pop = self.peek(stack_id)
+
+        # special case, when there is only one element in this stack,
+        # we will clear top index, and reset next to sentinel value (-1)
+
+        if self.get_size(stack_id) == 1:
```

Comment:
is there a way in which this special case can be avoided?

Reply:
This code was brought in based on the idea of "resetting" stack state which doesn't go back into its original "new" state.  That is, setting the next index as the sentinel value (-1)....Even though it wouldn't contribute to the correctness of the other functions. 

For an empty stack, it doesn't matter what the starting value of `top_index_next` is because when we push an element onto an empty stack, the `top_index_next` will be created based off the value of `top_index`.  So, all that matters is that `top_index` is reset correctly, for both `top_index_next` and `top_index` to work.

This means that the code I added was redundant.  My thinking was too narrow while writing that.
The solution will be to remove code related to resetting `top_index_next`.

Reply:
FIxed here:  7a4d520912577a65200baa9f5670a6c3dc698edd